src/mock_parsec/state.rs

```diff
@@ -0,0 +1,77 @@
+// Copyright 2019 MaidSafe.net limited.
+//
+// This SAFE Network Software is licensed to you under The General Public License (GPL), version 3.
+// Unless required by applicable law or agreed to in writing, the SAFE Network Software distributed
+// under the GPL Licence is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied. Please review the Licences for the specific language governing
+// permissions and limitations relating to use of the SAFE Network Software.
+
+//! Shared state for all mock parsec instances within a single test.
+
+use super::{
+    observation::{ObservationHolder, ObservationState},
+    Block, NetworkEvent, PublicId,
+};
+use crate::sha3::Digest256;
+use std::{
+    any::Any,
+    cell::RefCell,
+    collections::{BTreeMap, HashMap},
+};
+
+pub(super) struct SectionState<T: NetworkEvent, P: PublicId> {
+    pub observations: BTreeMap<ObservationHolder<T, P>, ObservationState<P>>,
+    pub blocks: Vec<(Block<T, P>, ObservationHolder<T, P>)>,
+}
+
+impl<T: NetworkEvent, P: PublicId> SectionState<T, P> {
+    fn new() -> Self {
+        Self {
+            observations: BTreeMap::new(),
+            blocks: Vec::new(),
+        }
+    }
+}
+
+type NetworkState<T, P> = HashMap<Digest256, SectionState<T, P>>;
+
+thread_local! {
+    static STATE: RefCell<Option<Box<dyn Any>>> = RefCell::new(None);
```

Comment:
Why `Any` and not `NetworkState`?

Reply:
`NetworkState` is generic. I'd have to instantiate it with concrete types before putting it into `thread_local` which I don't think I can, because at least the `P` param is different depending on whether we use mock crypto or not. Alternative would be to use conditional compilation. Not sure which is worse :thinking: